### PR TITLE
Update transient cache for `wp_add_global_styles_for_blocks`

### PR DESCRIPTION
**Pull request by joemcgill**
This uses a consistent cache key with the cache busting hash saved to the cache value rather than as part of the key. This allows us to remove the expiry time for the cache and only invalidate when the hash changes.

This also swaps the use of a site transient to the use of a regular transient so that the option will be autoloaded since there is no expiration time.

Trac ticket: https://core.trac.wordpress.org/ticket/61679

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/global-styles-and-settings.php [hunk 1]
```diff
@@ -320,10 +321,13 @@ function wp_add_global_styles_for_blocks() {
 			)
 		);
 
-		$cache_key = "wp_styles_for_blocks:$hash";
-		$cached    = get_site_transient( $cache_key );
-		if ( ! is_array( $cached ) ) {
-			$cached = array();
+		$cache_key = 'wp_styles_for_blocks';
+		$cached    = get_transient( $cache_key );
```
Comment (spacedmonkey) [inline src/wp-includes/global-styles-and-settings.php]: Why is this changed to `get_transient ` from `get_site_transient`?

**Review by spacedmonkey — COMMENTED** [review]
One question. But what worries me more is that unit tests are currently failing.

File: src/wp-includes/global-styles-and-settings.php (hunk 1 above)
Comment (joemcgill) [reply src/wp-includes/global-styles-and-settings.php]: This cache needs to be site specific, because it uses global settings and the computed styles for blocks in the hash, which incorporates users settings. When the hash key is part of the cache key name, every site on a multisite network with different settings will generate unique keys, which will expire naturally in 10 minutes.

To avoid the extra DB lookup for a transient with an expiration, I'm looking at making the cache key consistent, and put the hash inside the cached data for invalidation, but that means we need to save separate transients for each site on a multisite install, rather than as a shared transient for the whole network.

**Review by joemcgill — COMMENTED** [review]
Thanks @spacedmonkey. The unit test failures are expected currently, because I didn't update the new tests yet. I wanted to get feedback on the approach first, but will update the test before making this change (if at all).

**Comment by spacedmonkey** [comment]
@joemcgill On a side now, `get_site_transient` only result in one database query now. See

Screenshot 2024-06-25 at 22 52 24

Same with `get_transient`

Screenshot 2024-06-25 at 22 53 38

File: src/wp-includes/global-styles-and-settings.php [hunk 2]
```diff
@@ -257,21 +257,17 @@ function wp_add_global_styles_for_blocks() {
...
-					'block_nodes'    => $block_nodes,
-				)
-			)
-		);
-
-		$cache_key = "wp_styles_for_blocks:$hash";
-		$cached    = get_site_transient( $cache_key );
-		if ( ! is_array( $cached ) ) {
-			$cached = array();
+		// Hash the merged WP_Theme_JSON data to bust cache is settings or styles change.
```
Comment (costdev) [inline src/wp-includes/global-styles-and-settings.php]:
```suggestion
		// Hash the merged WP_Theme_JSON data to bust cache on settings or styles change.
```

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php [hunk 3]
```diff
@@ -334,25 +331,4 @@ private function get_global_styles() {
 		$actual = wp_styles()->get_data( 'global-styles', 'after' );
 		return is_array( $actual ) ? $actual : array();
 	}
-
-	/**
-	 * Get cache key for `wp_styles_for_blocks`.
-	 *
-	 * @return string The cache key.
-	 */
-	private function get_wp_styles_for_blocks_cache_key() {
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]: Curious, why remove this test rather than adjust it?

Comment (joemcgill) [reply tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]: It's no longer needed after the changes. This is a utility method to make it easier to generate the previous cache key value that is no longer needed.

File: src/wp-includes/global-styles-and-settings.php [hunk 4]
```diff
@@ -257,21 +257,17 @@ function wp_add_global_styles_for_blocks() {
...
-				)
-			)
-		);
-
-		$cache_key = "wp_styles_for_blocks:$hash";
-		$cached    = get_site_transient( $cache_key );
-		if ( ! is_array( $cached ) ) {
-			$cached = array();
+		// Hash the merged WP_Theme_JSON data to bust cache on settings or styles change.
+		$cache_hash = md5( wp_json_encode( $tree->get_raw_data() ) );
```
Comment (joemcgill) [inline src/wp-includes/global-styles-and-settings.php:266]: This is a much simpler hash generation, and accounts for a much broader set of changes that might happen to the merged Theme JSON data, which could affect global block styles, which is less risky.

**Review by aaronrobertshaw — COMMENTED** [review]
> This has been updated to also fix a bug that was reported in https://core.trac.wordpress.org/ticket/59595#comment:71 with testing instructions.

Thanks @joemcgill for updating this PR with a fix for the [reported issue](https://core.trac.wordpress.org/ticket/59595#comment:71) 👍 

After applying the changes here, the global block styles update as expected on the frontend.

| Before Revert (ad135d5) | After |
|---|---|
| <video src="https://github.com/user-attachments/assets/3770f7dc-c3ba-485a-a549-cd558a13b8e6" /> | <video src="https://github.com/user-attachments/assets/b1ea2128-18e5-4e99-8df7-61494199b662" /> |

It looks like this PR needs a rebase but in general it's looking pretty close to me.

**Review by andrewserong — COMMENTED** [review]
> but in general it's looking pretty close to me.

Same here, I like how this makes for a simpler and broader hash, it sounds like this will be a fair bit safer 👍

**Comment by joemcgill** [comment]
Thanks all, I'll update this PR so it applies cleanly after the revert of the original caching so we can try again for 6.6.1

**Comment by joemcgill** [comment]
I've refreshed this PR against trunk, which adds back the caching that was reverted in https://core.trac.wordpress.org/changeset/58710 using the updated approach.

I spent some time trying to write a new PHPUnit test that will confirm that the block meta will change whenever the `wp_global_styles` post is updated but discovered that the query in `WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( wp_get_theme() )` does not return the same post that is created if you first call `WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( wp_get_theme(), true )` to create the post. This seems like an unrelated bug so will report it upstream, but wanted to mention in case anyone had any insight into why this might be expected behavior.

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php [hunk 5]
```diff
@@ -75,6 +75,84 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
...
+	 * @ticket 59595
+	 */
+	public function test_styles_for_blocks_cache_is_set() {
+		$this->set_up_third_party_block();
+
+		wp_register_style( 'global-styles', false, array(), true, true );
+
+		$cache_key                = 'wp_styles_for_blocks';
+		$styles_for_blocks_before = get_transient( $cache_key );
+		$this->assertFalse( $styles_for_blocks_before );
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]:
```suggestion
		$this->assertFalse( $styles_for_blocks_before, 'No block styles should be cached yet.' );
```
For tests with multiple assertions, they should have a final `$message` argument to add context to the failure, even if the assertion methods are different. This is really useful for debugging because it immediately shows what was being tested at that point in the test, and makes reading multiple test failures easier. [Ref](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions)

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php [hunk 6]
```diff
@@ -75,6 +75,84 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
...
+
+		wp_add_global_styles_for_blocks();
+
+		$styles_for_blocks_after = get_transient( $cache_key );
+		$this->assertNotEmpty( $styles_for_blocks_after );
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]:
```suggestion
		$this->assertNotEmpty( $styles_for_blocks_after, 'No block styles were cached.' );
```
See above

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php [hunk 7]
```diff
@@ -75,6 +75,84 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
...
+		/*
+		 * Call register of global styles again to ensure the cache is updated.
+		 * In normal conditions, this function is only called once per request.
+		 */
+		wp_add_global_styles_for_blocks();
+
+		$styles_for_blocks_updated = get_transient( 'wp_styles_for_blocks' );
+		$this->assertNotEmpty( $styles_for_blocks_updated, 'Updated cache was not set.' );
+
+		$this->assertNotEquals(
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]: Does this fail with `assertNotSame()`? If not, then `assertSame()` should be used since we always use the strictest assertion.

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php [hunk 8]
```diff
@@ -75,6 +75,84 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
 		);
 	}
 
+	/**
+	 * Ensure that the block cache is set for global styles.
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]:
```suggestion
	 * Tests that the block cache is set for global styles.
```
Nit: Tests usually begin with "Tests that", and this aligns with the use of third-party singular verbs used in method docblocks. [Ref](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#documenting-tips)

:1234: Applies to other tests in the PR.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the work on this @joemcgill! The current  tests look good. I've added some standards suggestions in this review.

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php (hunk 7 above)
Comment (joemcgill) [reply tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]: Fixed in [683d07a](https://github.com/WordPress/wordpress-develop/pull/6879/commits/683d07ae9981eda30cba2a51caaebffa6b82f0c6)

File: tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php (hunk 8 above)
Comment (joemcgill) [reply tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php]: Happy to make that change to the included tests. Also updated the ticket number now that this is being associated by a new Trac ticket.

See: [6c35708](https://github.com/WordPress/wordpress-develop/pull/6879/commits/6c35708ef008b468b9afac9e2648ebc5d9af71dc)

**Comment by joemcgill** [comment]
Thanks for the review, @costdev. This is ready for another look.

**Comment by joemcgill** [comment]
> I spent some time trying to write a new PHPUnit test that will confirm that the block meta will change whenever the wp_global_styles post is updated but discovered that the query in WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( wp_get_theme() ) does not return the same post that is created if you first call WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles( wp_get_theme(), true ) to create the post. This seems like an unrelated bug so will report it upstream, but wanted to mention in case anyone had any insight into why this might be expected behavior.

[Gutenberg issue opened](https://github.com/WordPress/gutenberg/issues/63722), along with [a unit test](https://github.com/WordPress/gutenberg/pull/63721) that demonstrates the behavior.

**Review by costdev — APPROVED** [review]
Thanks for the updates @joemcgill! Looks good to me :+1:

File: src/wp-includes/global-styles-and-settings.php [hunk 9]
```diff
@@ -255,8 +255,46 @@ function wp_add_global_styles_for_blocks() {
...
-		$block_css = $tree->get_styles_for_block( $metadata );
+
+		if ( $can_use_cached ) {
+			// Use the block name as the key for cached CSS data. Otherwise, use a hash of the metadata.
+			$cache_node_key = isset( $metadata['name'] ) ? $metadata['name'] : md5( wp_json_encode( $metadata ) );
+
+			if ( isset( $cached['blocks'][ $cache_node_key ] ) ) {
+				$block_css = $cached['blocks'][ $cache_node_key ];
+			} else {
+				$block_css                           = $tree->get_styles_for_block( $metadata );
```
Comment (peterwilsoncc) [inline src/wp-includes/global-styles-and-settings.php:291]: Just to make sure, can the meta data include block settings in the editor that will effect the output styles? 

If so, then this may have incorrect data when a block is updated. If not then it's dandy.

Comment (aaronrobertshaw) [reply src/wp-includes/global-styles-and-settings.php:291]: I'm not sure I fully grasp the question, sorry.

Global styles for blocks set in the editor are saved to the Global Styles CPT, which are retrieved and merged with base and theme data via `WP_Theme_JSON_Resolver::get_merged_data();` above. While in the editor, if global styles for a block type are adjusted without saving, there's a JS hook that outputs updated styles overriding those enqueued here. 

My understanding is that the styles generated here only need to reflect what is saved for Global Styles.

Not sure if that helps, answers the question, or just confuses things 😅

Comment (peterwilsoncc) [reply src/wp-includes/global-styles-and-settings.php:291]: > My understanding is that the styles generated here only need to reflect what is saved for Global Styles.

That helps and was what I was asking, thanks Aaron.

**Comment by joemcgill** [comment]
I've refreshed this PR and added an additional unit test to confirm that the cache gets invalidated whenever the global-styles post is updated in the editor.

**Comment by joemcgill** [comment]
I've run a set of benchmarks on this locally, which compares 40 runs to the homepage of Twenty Twenty Four. 

| URL                                 | trunk                  | This PR                | Difference             | Percentage Difference  |
|-------------------------------------|------------------------|------------------------|------------------------|------------------------|
| Response Time (p10)                 | 70.75                  | 69.21                  | -1.54                  | -2.22%                 |
| Response Time (p25)                 | 71.51                  | 69.77                  | -1.74                  | -2.50%                 |
| Response Time (p50)                 | 72.75                  | 70.65                  | -2.10                  | -2.97%                 |
| Response Time (p75)                 | 74.45                  | 72.05                  | -2.40                  | -3.33%                 |
| Response Time (p90)                 | 76.19                  | 76.28                  | 0.09                   | 0.12%                  |
| wp-before-template (p10)            | 31.48                  | 31.21                  | -0.27                  | -0.86%                 |
| wp-before-template (p25)            | 31.83                  | 31.54                  | -0.29                  | -0.92%                 |
| wp-before-template (p50)            | 32.25                  | 32.17                  | -0.08                  | -0.25%                 |
| wp-before-template (p75)            | 33.25                  | 33.13                  | -0.12                  | -0.36%                 |
| wp-before-template (p90)            | 33.8                   | 34.42                  | 0.62                   | 1.80%                  |
| wp-template (p10)                   | 36.39                  | 35.08                  | -1.31                  | -3.73%                 |
| wp-template (p25)                   | 36.91                  | 35.43                  | -1.48                  | -4.18%                 |
| wp-template (p50)                   | 37.83                  | 35.96                  | -1.87                  | -5.20%                 |
| wp-template (p75)                   | 39.14                  | 37.75                  | -1.39                  | -3.55%                 |
| wp-template (p90)                   | 40.15                  | 39.07                  | -1.08                  | -2.69%                 |
| wp-total (p10)                      | 68.47                  | 66.85                  | -1.62                  | -2.42%                 |
| wp-total (p25)                      | 69.49                  | 67.45                  | -2.04                  | -3.02%                 |
| wp-total (p50)                      | 70.48                  | 68.3                   | -2.18                  | -3.19%                 |
| wp-total (p75)                      | 71.56                  | 69.64                  | -1.92                  | -2.68%                 |
| wp-total (p90)                      | 73.91                  | 74.3                   | 0.39                   | 0.53%                  |

**Comment by joemcgill** [comment]
Merged in https://core.trac.wordpress.org/changeset/59256
